### PR TITLE
v1/stream/query interface subscription support

### PR DESCRIPTION
**Pull request by chunlokling-da**
Fixes https://github.com/digital-asset/daml/issues/14817

```rest
CHANGELOG_BEGIN
Users can now subscribe `v1/stream/query` with queries with single interface id.
CHANGELOG_END
```

### CODE REVIEW

File: ledger-service/http-json/src/main/scala/com/digitalasset/http/WebSocketService.scala [hunk 1]
```diff
@@ -799,98 +868,118 @@ class WebSocketService(
...
+                    jwt,
+                    ledgerId,
+                    parties,
+                    resolvedQuery.resolved.toList,
+                    liveStartingOffset,
+                    Terminates.Never,
+                  )
+                  .via(
+                    convertFilterContracts(
+                      domain.ResolvedQuery.ContractTypeIdsQuery(resolvedQuery.resolved),
```
Comment (ray-roestenburg-da) [inline ledger-service/http-json/src/main/scala/com/digitalasset/http/WebSocketService.scala:939]: Can't you just use `resolvedQuery` here?

**Comment by chunlokling-da** [comment]
Spin off https://github.com/digital-asset/daml/issues/14931

**Review by ray-roestenburg-da — COMMENTED** [review]
Please add a test to verify that the interface view returned has the isAbcPrefix and is123Suffix fields, otherwise LGTM

**Review by ray-roestenburg-da — APPROVED** [review]

**Comment by ray-roestenburg-da** [comment]
@chunlokling-da let's have @S11001001 take another look at this tomorrow before merging?

**Comment by chunlokling-da** [comment]
Yeah sure!

File: ledger-service/http-json/src/main/scala/com/digitalasset/http/WebSocketService.scala [hunk 2]
```diff
@@ -63,14 +66,23 @@ object WebSocketService {
...
-      fn: (domain.ActiveContract[LfV], Option[domain.Offset]) => Option[Positive],
-      dbQuery: (domain.PartySet, dbbackend.ContractDao) => ConnectionIO[
-        _ <: Vector[(domain.ActiveContract[JsValue], Positive)]
-      ],
-  )
+  private sealed abstract class StreamPredicate[+Positive] extends Product with Serializable
+
+  private object StreamPredicate {
+    final case class ValidStreamPredicate[+Positive](
+        resolvedQurey: domain.ResolvedQuery,
```
Comment (S11001001) [inline ledger-service/http-json/src/main/scala/com/digitalasset/http/WebSocketService.scala:73]:
```suggestion
    final case class Valid[+Positive](
        resolvedQuery: domain.ResolvedQuery,
```

File: ledger-service/http-json/src/main/scala/com/digitalasset/http/WebSocketService.scala [hunk 3]
```diff
@@ -63,14 +66,23 @@ object WebSocketService {
...
+        fn: (domain.ActiveContract[LfV], Option[domain.Offset]) => Option[Positive],
+        dbQuery: (domain.PartySet, dbbackend.ContractDao) => ConnectionIO[
+          _ <: Vector[(domain.ActiveContract[JsValue], Positive)]
+        ],
+    ) extends StreamPredicate[Positive]
+    final case class AllContractTypeIdsNotResolved[+Positive](
+        unresolved: NonEmpty[Set[domain.TemplateId.OptionalPkg]]
+    ) extends StreamPredicate[Positive]
+    final case class UnsupportedQuery[+Positive](reason: Unsupported)
+        extends StreamPredicate[Positive]
```
Comment (S11001001) [inline ledger-service/http-json/src/main/scala/com/digitalasset/http/WebSocketService.scala:84]:
```suggestion
    final case class AllContractTypeIdsNotResolved(
        unresolved: NonEmpty[Set[domain.TemplateId.OptionalPkg]]
    ) extends StreamPredicate[Nothing]
    final case class UnsupportedQuery(reason: Unsupported)
        extends StreamPredicate[Nothing]
```

**Review by S11001001 — APPROVED** [review]
Looks mostly good to me, but please address these before merging.

File: ledger-service/fetch-contracts/src/main/scala/fetchcontracts/domain/ContractTypeId.scala [hunk 4]
```diff
@@ -124,6 +123,35 @@ object ResolvedQuery {
     def resolved: Set[ContractTypeId.Resolved] =
       Set(interfaceId).toSet[ContractTypeId.Resolved]
   }
+
+  implicit def `Unsupported or ResolvedQuery monoid`: Monoid[Unsupported \/ ResolvedQuery] = {
```
[the remark this answers is not in the document]
Comment (chunlokling-da) [reply ledger-service/fetch-contracts/src/main/scala/fetchcontracts/domain/ContractTypeId.scala:125]: Thanks for pointing this out. Do you mean by creating a new type like a wrapper? 
I am thinking to define a private (or local?) wrapper class and the Monoid.
```scala
  private case class UnsupportedOrResolvedQuery(value: Unsupported \/ ResolvedQuery)
  private implicit val `Unsupported or ResolvedQuery monoid`: Monoid[UnsupportedOrResolvedQuery] = {...}
```

[the remark this answers is not in the document]
Comment (chunlokling-da) [reply ledger-service/fetch-contracts/src/main/scala/fetchcontracts/domain/ContractTypeId.scala:125]: As discussed with @S11001001 , I will tag Unsupporged \/ ResolvedQuery 
```
private sealed trait UnsupportedOrResolvedQueryTag
Monoid[Unsupported \/ ResolvedQuery @@ UnsupportedOrResolvedQueryTag]
```

File: ledger-service/http-json/src/main/scala/com/digitalasset/http/WebSocketService.scala [hunk 5]
```diff
@@ -320,39 +332,64 @@ object WebSocketService {
...
-              .map(_ map (_ rightMap (_ map posMap)))
-          },
+          (unsupportedOrResolvedQuery, resolved, unresolved, q) = res
+        } yield (
+          unresolved match {
+            case NonEmpty(unresolvedSet) if resolved.isEmpty =>
+              StreamPredicate.AllContractTypeIdsNotResolved(unresolvedSet)
+            case _ =>
+              unsupportedOrResolvedQuery match {
+                case \/-(resolvedQuery: ResolvedQuery) =>
```
[the remark this answers is not in the document]
Comment (S11001001) [reply ledger-service/http-json/src/main/scala/com/digitalasset/http/WebSocketService.scala]: This was not addressed before merge.

[the remark this answers is not in the document]
Comment (S11001001) [reply ledger-service/http-json/src/main/scala/com/digitalasset/http/WebSocketService.scala]: Rather, it was addressed here, but...

File: ledger-service/http-json/src/main/scala/com/digitalasset/http/WebSocketService.scala [hunk 6]
```diff
@@ -525,7 +602,18 @@ object WebSocketService {
         )
         .map { case (resolvedWithKey, unresolved) =>
           val q = getQ(resolvedWithKey)
-          streamPredicate(q, unresolved)
+          unresolved match {
+            case NonEmpty(unresolvedSet) if q.isEmpty =>
+              StreamPredicate.AllContractTypeIdsNotResolved(unresolvedSet)
+            case _ =>
+              domain.ResolvedQuery(q.keySet) match {
+                case \/-(resolvedQuery: ResolvedQuery) =>
```
Comment (S11001001) [inline ledger-service/http-json/src/main/scala/com/digitalasset/http/WebSocketService.scala:610]: ...this one was not addressed.

Comment (S11001001) [reply ledger-service/http-json/src/main/scala/com/digitalasset/http/WebSocketService.scala:610]: #14966

Comment (chunlokling-da) [reply ledger-service/http-json/src/main/scala/com/digitalasset/http/WebSocketService.scala:610]: oh sorry

**Pull request merged**